src/DefaultBuilder/src/WebApplication.cs

```diff
@@ -237,4 +241,39 @@ private void Listen(string? url)
         addresses.Clear();
         addresses.Add(url);
     }
+
+    private string DebuggerToString()
+    {
+        return $@"ApplicationName = ""{Environment.ApplicationName}"", IsRunning = {(IsRunning ? "true" : "false")}";
+    }
+
+    // Web app is running if the app has been started and hasn't been stopped.
+    private bool IsRunning => Lifetime.ApplicationStarted.IsCancellationRequested && !Lifetime.ApplicationStopped.IsCancellationRequested;
+
+    internal sealed class WebApplicationDebugView(WebApplication webApplication)
+    {
+        private readonly WebApplication _webApplication = webApplication;
+
+        public IServiceProvider Services => _webApplication.Services;
+        public IConfiguration Configuration => _webApplication.Configuration;
+        public IWebHostEnvironment Environment => _webApplication.Environment;
+        public IHostApplicationLifetime Lifetime => _webApplication.Lifetime;
+        public ILogger Logger => _webApplication.Logger;
+        public string Urls => string.Join(", ", _webApplication.Urls);
+        public IList<Endpoint> Endpoints => _webApplication.DataSources.SelectMany(ds => ds.Endpoints).ToList();
```

Comment:
If the goal is to show a flattened list of all the endpoints, and not an exact representation of `_webApplication.DataSources`, it'd be better to use the root CompositeEndpointData source which gets created from `RouteOptions.EndpointDataSources`, includes datasources added inside of custom calls to `UseEndpoints`, and is used for link generation.

```suggestion
        public IList<Endpoint> Endpoints => _webApplication.Services.GetRequiredService<EndpointDataSource>().Endpoints;
```